base/strings/basic.jl

```diff
@@ -613,6 +613,33 @@ isascii(c::Char) = bswap(reinterpret(UInt32, c)) < 0x80
 isascii(s::AbstractString) = all(isascii, s)
 isascii(c::AbstractChar) = UInt32(c) < 0x80
 
+@inline function _isascii(code_units::AbstractVector{CU}, first, last) where {CU}
+    r = zero(CU)
+    for n = first:last
+        @inbounds r |= code_units[n]
+    end
+    return 0 ≤ r < 0x80
+end
+
+@inline _isascii(::Val{N},cu::AbstractVector{CU}, first) where {N,CU} = @inline _isascii(cu,first,first+N-1)
+
+#The chunking algorithm makes the last two chunks overlap inorder to keep the size fixed
+@inline function  _isascii_chunks(::Val{N},cu::AbstractVector{CU}, first,last) where {N,CU}
+    for n=first:N:last - N
+        _isascii(Val(N),cu,n) || return false
+   end
+   return  _isascii(Val(N),cu,last-N+1)
+end
```

Comment:
```suggestion
@inline function  _isascii_chunks(::Val{N},cu::AbstractVector{CU}, start, stop) where {N,CU}
    ra = start:N:stop-N
    for n=ra
        _isascii(Val(N),cu,n) || return false
    end
    return  _isascii(cu,last(ra)+1, stop)
end
```

I still wonder whether something like this is worthwhile :thinking: I had some old experiments somewhere in the process where a version like this:

```julia
@inline function  _isascii_chunks(::Val{N},cu::AbstractVector{CU}, start, stop) where {N,CU}
   ra = start:N:stop-N
   let ra = ra
   for n=ra
        _isascii(Val(N),cu,n) || return false
   end
   end
   return  _isascii(cu,last(ra)+1, stop)
end
```

was oddly enough faster than both of the above, but I don't have any benchmarks at hand to back that up :/

Reply:
It'd be surprising to me if that is what makes it fast, since it's doing potentially a lot of work for nothing :thinking: I guess it allows to always take the SIMD path..

The thinking with `let` was that it introduces a hard scope, in which it's known that `ra` is only used once. This disentangles the object from the `last` use later on, since the `ra` in the `let` block is technically speaking a different object. At the same time, the compiler is smart enough to infer that the only use of `ra` outside of the `let` is to compute the `last` element, which it needs to do anyway during construction of the object. All other parts of the object (i.e. start and step) aren't loaded at all, so can be eliminated entirely.